src/Mvc/test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/HtmlHelperComponentsExtensionsTests.cs

```diff
@@ -0,0 +1,330 @@
+                    TemperatureC = -13,
+                    Summary = "Freezing",
+                    TemperatureF = 9
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "09/05/2018",
+                    TemperatureC = -16,
+                    Summary = "Balmy",
+                    TemperatureF = 4
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "10/05/2018",
+                    TemperatureC = 2,
+                    Summary = "Chilly",
+                    TemperatureF = 29
+                }
+            };
+
+            public class WeatherRow
+            {
+                public string DateFormatted { get; set; }
+                public int TemperatureC { get; set; }
+                public string Summary { get; set; }
+                public int TemperatureF { get; set; }
+            }
+
+            public WeatherRow[] RowsToDisplay { get; set; }
+
+            protected override async Task OnParametersSetAsync()
+            {
+                // Simulate an async workflow.
+                await Task.Yield();
+                RowsToDisplay = _weatherData;
+            }
+
+            protected override void BuildRenderTree(RenderTreeBuilder builder)
+            {
+                base.BuildRenderTree(builder);
+                var s = 0;
+                builder.OpenElement(s++, "table");
+                builder.AddMarkupContent(s++, "\n");
+                builder.OpenElement(s++, "thead");
+                builder.AddMarkupContent(s++, "\n");
+                builder.OpenElement(s++, "tr");
+                builder.AddMarkupContent(s++, "\n");
+
+                builder.OpenElement(s++, "th");
+                builder.AddContent(s++, "Date");
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+
+                builder.OpenElement(s++, "th");
+                builder.AddContent(s++, "Summary");
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+
+                builder.OpenElement(s++, "th");
+                builder.AddContent(s++, "F");
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+
+                builder.OpenElement(s++, "th");
+                builder.AddContent(s++, "C");
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+                builder.CloseElement();
+                builder.AddMarkupContent(s++, "\n");
+                builder.OpenElement(s++, "tbody");
+                builder.AddMarkupContent(s++, "\n");
+                if (RowsToDisplay != null)
+                {
+
+                    foreach (var element in RowsToDisplay)
+                    {
+
+                        builder.OpenElement(s++, "tr");
```

Comment:
If you want the diffing algorithm to behave normally, you'll need to reset the sequence number here back to the same value at the start of each loop cycle.

Reply:
https://github.com/aspnet/AspNetCore/pull/6727/files#diff-1764a91a64f023c7b273a6219239f3efR274